torch/functional.py

```diff
@@ -543,3 +544,241 @@ def argsort(input, dim=None, descending=False):
+            'fro'  Frobenius norm                --
+            'nuc'  nuclear norm                  --
+            inf    max(sum(abs(x), dim=1))       max(abs(x))
+            -inf   min(sum(abs(x), dim=1))       min(abs(x))
+            0      --                            sum(x != 0)
+            1      max(sum(abs(x), dim=0))       as below
+            -1     min(sum(abs(x), dim=0))       as below
+            2      largest singular value        as below
+            -2     smallest singular value       as below
+            other  as vec norm when dim is None  sum(abs(x)**ord)**(1./ord)
+            =====  ============================  ==========================
+        dim ({int, 2-tuple of ints, 2-list of ints}): If it is a int, vector norm
+        will be calculated, if it is 2-tuple of ints, matrix norm will be
+        calculated.
+        keepdim (bool): whether the output tensors have :attr:`dim`
+            retained or not. Ignored if ``dim=None``.
+        out (Tensor, optional) – the output tensor
+
+    Example::
+        >>> import torch
+        >>> a = torch.arange(9, dtype= torch.float) - 4
+        >>> b = a.reshape((3, 3))
+        >>> torch.norm(a)
+        tensor(7.7460)
+        >>> torch.norm(b)
+        tensor(7.7460)
+        >>> torch.norm(a, float('inf'))
+        tensor(4.)
+        >>> torch.norm(b, float('inf'))
+        tensor(9.)
+        >>> torch.norm(a, float('-inf'))
+        tensor(0.)
+        >>> torch.norm(b, float('-inf'))
+        tensor(2.)
+        >>> torch.norm(a, 1)
+        tensor(20.)
+        >>> torch.norm(b, 1)
+        tensor(20.)
+        >>> torch.norm(a, -1)
+        tensor(0.)
+        >>> torch.norm(b, -1)
+        tensor(0.)
+        >>> torch.norm(a, 2)
+        tensor(7.7460)
+        >>> torch.norm(b, 2)
+        tensor(7.7460)
+        >>> torch.norm(a, -2)
+        tensor(0.)
+        >>> torch.norm(b, -2)
+        tensor(0.)
+        >>> torch.norm(a, 3)
+        tensor(5.8480)
+        >>> torch.norm(a, -3)
+        tensor(0.)
+        >>> c = torch.tensor([[ 1, 2, 3],[-1, 1, 4]] , dtype= torch.float)
+        >>> torch.norm(c, dim=0)
+        tensor([1.4142, 2.2361, 5.0000])
+        >>> torch.norm(c, dim=1)
+        tensor([3.7417, 4.2426])
+        >>> torch.norm(c, p=1, dim=1)
+        tensor([6., 6.])
+        >>> d = torch.arange(8, dtype= torch.float).reshape(2,2,2)
+        >>> torch.norm(d, dim=(1,2))
+        tensor([ 3.7417, 11.2250])
+        >>> torch.norm(d[0, :, :]), torch.norm(d[1, :, :])
+        (tensor(3.7417), tensor(11.2250))
+    """
+    ndim = input.dim()
+
+    # catch default case
+    if dim is None and out is None:
+        if p == "fro":
+            return torch._C._VariableFunctions.norm(input, 2)
+        if isinstance(p, int):
+            return torch._C._VariableFunctions.norm(input, p)
+        if isinstance(p, float) and p != inf and p != -inf:
+            return torch._C._VariableFunctions.norm(input, p)
```

Comment:
I think this is a bit too complex. My original idea is that most of the code should live in ATen. including the arg checking code below, with only minimal Python code here doing the actual dispatch. E.g., the entire Python function body may just look like 
```py
if p == 'nuc':
  return torch._C._VariableFunctions.nuc_norm(...)
elif p == 'fro'
  return torch._C._VariableFunctions.fro_norm(...)
else:
  return torch._C._VariableFunctions.norm(...)
```

Reply:
@ezyang when it comes to computing arbitrary norms (or distances) it seems like the appropriate c++ style is just to have different names for the functions i.e. how this PR is structured. Given that absent a reason for a better api torch should probably follow numpy's i.e. lgtm!

I will second @SsnL that python should probably just convert strings to appropriate c++ calls, maybe with some extra checking for places where python doesn't enforce the same typing that c++ does.

Reply:
@erikbrinkman That is the current way. However, in the long term goal, we want to have enum support in native function, and to allow specifying how to map python objects to enum values. That will make this, nn.functional.pad, and many other things much easier to write!